src/lib.rs

```diff
@@ -2,269 +2,266 @@
-    let target = target.as_ref();
-    match kind {
-        // Single files
-        Kind::BZip2 => source.bz2().file(target),
-        Kind::GZip => source.gz().file(target),
-        Kind::LZMA => source.lzma().file(target),
-        Kind::LZip => source.lzip().file(target),
-        Kind::Xz => source.xz().file(target),
-
-        // Archive files
-        Kind::Tar => source.tar(target),
-        Kind::TarBZip2 => source.bz2().tar(target),
-        Kind::TarGZip => source.gz().tar(target),
-        Kind::TarLZMA => source.lzma().tar(target),
-        Kind::TarLZip => source.lzip().tar(target),
-        Kind::TarXz => source.xz().tar(target),
-        Kind::Zip => source.unzip(target),
-    }
+    run_with_archive(Mode::Raw, source, |ar, _aw, mut entry| unsafe {
+        archive_result(ffi::archive_read_next_header(ar, &mut entry), ar)?;
+        write_data_block(target, ar)
+    })
 }
 
-#[cfg(test)]
-mod test {
-    use super::*;
-    use std::{
-        env, fs,
-        io::{Read, Write},
-        os::unix::fs::PermissionsExt,
-        path::PathBuf,
-    };
-    use tempfile::TempDir;
+/// Reads the input buffer and extract to directory.
+///
+/// # Arguments
+///
+/// * `source` - Reader.
+/// * `path` - Path to extract the target buffer.
+///
+/// # Example
+///
+/// uncompress_archive(std::io::Read, &str)?;
+pub fn uncompress_archive<R>(source: &mut R, dir_path: &Path) -> Result<()>
+where
+    R: Read + 'static,
+{
+    run_with_archive(Mode::Disk, source, |ar, aw, mut entry| unsafe {
+        loop {
+            match ffi::archive_read_next_header(ar, &mut entry) {
+                ffi::ARCHIVE_EOF => return Ok(()),
+                ffi::ARCHIVE_OK => {
+                    let file_name = ffi::archive_entry_pathname(entry);
+                    let target_path = format!(
+                        "{}/{}",
+                        dir_path.to_str().unwrap(),
+                        CStr::from_ptr(file_name).to_str()?.to_string()
+                    );
+                    ffi::archive_entry_set_pathname(
+                        entry,
+                        CString::new(target_path).unwrap().into_raw(),
+                    );
+                    let link_name = ffi::archive_entry_hardlink(entry);
+                    if !link_name.is_null() {
+                        let target_path = format!(
+                            "{}/{}",
+                            dir_path.to_str().unwrap(),
+                            CStr::from_ptr(link_name).to_str()?.to_string()
+                        );
+                        ffi::archive_entry_set_hardlink(
+                            entry,
+                            CString::new(target_path).unwrap().into_raw(),
+                        );
+                    }
+                    ffi::archive_write_header(aw, entry);
+                    if ffi::archive_entry_size_is_set(entry) != ffi::ARCHIVE_OK
+                        || ffi::archive_entry_size(entry) > 0
```

Comment:
When is this condition false? Can it be removed?

Reply:
Yes, this can be removed